core/src/main/scala/org/apache/spark/SparkEnv.scala

```diff
@@ -355,6 +355,12 @@ object SparkEnv extends Logging {
 
     // Mapping from block manager id to the block manager's information.
     val blockManagerInfo = new concurrent.TrieMap[BlockManagerId, BlockManagerInfo]()
+    // Using a cache here since we only want to track recently removed executors to deny their
+    // block manager registration while their StopExecutor message is in-flight.
+    // Assuming average size of 6 bytes of execId and each entry in Cache taking around 64 bytes,
+    // max size of this cache = (6 + 64) * 30000 = 2.1MB
+    val recentlyRemovedExecutors = CacheBuilder.newBuilder().maximumSize(30000)
+      .build[String, String]()
```

Comment:
Any particular reason for such a high cache size ?

Reply:
The high cache size is to ensure the fix works for a large enough job with 30000 executors. 

Sure, does an expiry of 10min (or larger) sounds good?
This should give the executor long enough to process `StopExecutor` (in-flight) message and complete the shutdown.

Reply:
The timeout should be modeled based on what is the max expected delay for heartbeat to come in from executor.

Reply:
I think another possible solution is to extend the `BlockManagerInfo` with the timestamp of the removing. So modelling the removing as a new state and this way we could avoid using this separate cache completely and all the bm related data would be in the same place. 

Of course in this case you should implement the cleanup.

For example it could be just a simple Long var which is 0 by default which means the BlockManager is alive/active (this special value can be hidden behind a method of `BlockMangerInfo` like `isAlive(currentTs)`). The cleanup would triggered for delay plus some extra time to avoid too frequent iteration on the `blockManagerInfo` collection.

WDYT?

Reply:
I thought about this a bit more but haven't checked the code yet: Is it possible to separate driver commanded intentionally removed executors from unintentional executor loss?

Intentionally removed executors shouldn't be re-registered. 

cc @Ngone51

Reply:
I believe extending `BlockManagerInfo` should solve the problem.
However, we will have to abstract `blockManagerInfo: mutable.Map[BlockManagerId, BlockManagerInfo]`.
Currently, on `RemoveExecutor`, we remove the corresponding `BlockManagerInfo` from `blockManagerInfo` map, now since during the removal, instead of removing, we update the `removalTs` inside `BlockManagerInfo`, other methods will have to filter the values in the map first before using them.

So my suggestion here would be to abstract these details into a `BlockManagerEndpointSharedState` which holds the `blockManagerInfo` map and exposes methods for lookup and updates.
I propose the name as `BlockManagerEndpointSharedState` since we will have to pass the same object to `BlockManagerMasterEndpoint` and `BlockManagerMasterHeartbeatEndpoint`.

Reply:
When `CoarseGrainedSchedulerBackend` receives `RemoveExecutor`, it has the `ExecutorLossReason`. However, after processing the message, when it publishes `SparkListenerExecutorRemoved`, the reason is passed in form of a string.

Reply:
> Currently, on RemoveExecutor, we remove the corresponding BlockManagerInfo from blockManagerInfo map...

As per https://github.com/apache/spark/pull/32114#issuecomment-819255421, I don't think there would be a `BlockManagerMessages.RemoveExecutor` raised in this PR case.

Could you point out on which code path that the `BlockManagerMessages.RemoveExecutor` is raised? 

If there's no more code path raises the `BlockManagerMessages.RemoveExecutor` in this PR case, then @attilapiros definitely works. But, I'd also suggest another idea in  https://github.com/apache/spark/pull/32114#issuecomment-819255421.

Reply:
> Is it possible to separate driver commanded intentionally removed executors from unintentional executor loss?

@attilapiros  It's possible to know that an executor is removed intentionally by the driver or not. The problem is, currently, the executor info is stored in many different places. So you have to update many methods or messages to add the `isIntentional` filed (for example), to let all the components know and make certain decisions on it, which could be miscellaneous.

Reply:
Yes, @Ngone51 as pointed out in [#32114](https://github.com/apache/spark/pull/32114#issuecomment-819255421) `BlockManagerMessages.RemoveExecutor` in limited cases. I could not find any more code path apart from the one you pointed.

Reply:
I found a code path which will break @attilapiros's proposed solution since the following holds true.

> However, we will have to abstract blockManagerInfo: mutable.Map[BlockManagerId, BlockManagerInfo].
Currently, ...

Let's consider the following set of events:
- A  `CoarseGrainedClusterMessage.RemoveExecutor` is issued
- `CoarseGrainedSchedulerBackend` issues async `StopExecutor` on `executorEndpoint` and then invokes `executorLost` on `TaskSchedulerImpl`
https://github.com/apache/spark/blob/e60939591336a2c38e5b7a6a36f45b8614b46fe5/core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala#L430
- TaskSchedulerImpl in its `executorLost` invokes `dagScheduler.executorLost`
https://github.com/apache/spark/blob/e60939591336a2c38e5b7a6a36f45b8614b46fe5/core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala#L998-L1001
- `DAGScheduler` while handling executorLost invokes `removeExecutorAndUnregisterOutputs` which internally invokes `blockManagerMaster.removeExecutor(execId)` (as you pointed out in your [comment below](https://github.com/apache/spark/pull/32114#issuecomment-819255421)) which further clears `blockManagerId` from `blockManagerInfo` in `BlockManagerMasterEndpoint`
https://github.com/apache/spark/blob/e60939591336a2c38e5b7a6a36f45b8614b46fe5/core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala#L2043
- The Executor has not yet processed `StopExecutor`
- Executor reports its Heartbeat
- `HeartbeatReceiver` invokes `scheduler.executorHeartbeatReceived` to check if the BlockManager on the executor requires re-registration
https://github.com/apache/spark/blob/e60939591336a2c38e5b7a6a36f45b8614b46fe5/core/src/main/scala/org/apache/spark/HeartbeatReceiver.scala#L137
- `TashSchedulerImpl` delegates this to `DAGScheduler`
https://github.com/apache/spark/blob/e60939591336a2c38e5b7a6a36f45b8614b46fe5/core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala#L858
- `DAGScheduler` asks `BlockManagerMasterHeartbeatEndpoint` if it knows the BlockManager
https://github.com/apache/spark/blob/e60939591336a2c38e5b7a6a36f45b8614b46fe5/core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala#L300
- `BlockManagerMasterHeartbeatEndpoint` returns false since it cannot find `blockManagerId` in `BlockManagerInfo` indicating the blockManager should re-register
https://github.com/apache/spark/blob/e60939591336a2c38e5b7a6a36f45b8614b46fe5/core/src/main/scala/org/apache/spark/storage/BlockManagerMasterHeartbeatEndpoint.scala#L51
- BlockManager re-registers which publishes the `SparkListenerBlockManagerAdded` causing the inconsistent book-keeping in `AppStatusStore`
- Executor processes `StopExecutor` and exits.